curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1436,38 +1426,55 @@ CURVEFS_ERROR ChunkCacheManager::Flush(uint64_t inodeId, bool force,
-                << ",len:" << iter->second->GetLen() << ",inodeId:" << inodeId
-                << ",chunkIndex:" << index_;
-        assert(iter->second->IsDirty());
-        ret = iter->second->Flush(inodeId, force, toS3);
-        if ((ret != CURVEFS_ERROR::OK) && (ret != CURVEFS_ERROR::NOFLUSH)) {
-            LOG(WARNING) << "dataCache flush failed. ret:" << ret
-                         << ",index:" << index_
-                         << ",data chunkpos:" << iter->second->GetChunkPos();
-            return ret;
+    DataCachePtr dataCache;
+    while (1) {
+        bool isFlush = false;
+        {
+            WriteLockGuard writeLockGuard(rwLockChunk_);
+
+            auto iter = dataWCacheMap_.begin();
+            while (iter != dataWCacheMap_.end()) {
+                if (iter->second->CanFlush(force)) {
+                    dataCache = iter->second;
+                    dataWCacheMap_.erase(iter);
+                    isFlush = true;
+                    break;
+                } else {
+                    iter++;
+                }
+            }
         }
-        WriteLockGuard lockGuard(rwLockChunk_);
-        if (ret == CURVEFS_ERROR::OK) {
-            if (!iter->second->IsDirty()) {
+        if (isFlush) {
+            VLOG(9) << "Flush datacache chunkPos:" << dataCache->GetChunkPos()
+                    << ",len:" << dataCache->GetLen() << ",inodeId:" << inodeId
+                    << ",chunkIndex:" << index_;
+            assert(dataCache->IsDirty());
+            do {
+                ret = dataCache->Flush(inodeId, toS3);
+                if (ret == CURVEFS_ERROR::NOTEXIST) {
+                    LOG(WARNING)
+                        << "dataCache flush failed. ret:" << ret
+                        << ",index:" << index_
+                        << ",data chunkpos:" << dataCache->GetChunkPos();
+                    ReleaseWriteDataCache(dataCache);
+                    break;
+                } else if (ret == CURVEFS_ERROR::INTERNAL) {
+                    ::sleep(3);
+                    continue;
+                }
                 VLOG(9) << "ReleaseWriteDataCache chunkPos:"
-                        << iter->second->GetChunkPos()
-                        << ",len:" << iter->second->GetLen()
+                        << dataCache->GetChunkPos()
+                        << ",len:" << dataCache->GetLen()
                         << ",inodeId:" << inodeId << ",chunkIndex:" << index_;
                 if (!curvefs::client::common::FLAGS_enableCto) {
-                    AddReadDataCache(iter->second);
+                    WriteLockGuard lockGuard(rwLockChunk_);
+                    AddReadDataCache(dataCache);
                 }
-                ReleaseWriteDataCache(iter->second);
-            } else {
-                VLOG(6) << "data cache is dirty.";
-            }
+                ReleaseWriteDataCache(dataCache);
+            } while (ret != CURVEFS_ERROR::OK);
```

Comment:
you should reset `dataCache` to `nullptr` after `while`, because if dataCache is not null, and if you find another datacache that can be flushed in line 1438, the assignment operator will call previous datacache's destructor, and this happens when holding `rwLockChunk_`

Reply:
> you should reset `dataCache` to `nullptr` after `while`, because if dataCache is not null, and if you find another datacache that can be flushed in line 1438, the assignment operator will call previous datacache's destructor, and this happens when holding `rwLockChunk_`

fixed